sage.py

```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                    #print('---', att, [x[2] for x in episodes])
-                    keys.append(att)
-                    alerts.append(episodes)
-                else:
-                    c = 0
-                    ep = [x[2] for x in episodes]
-                    #print(ep)
-                    cuts = [i for i in range(len(episodes)-1) if (len(str(ep[i])) > len(str(ep[i+1]))) ]#(ep[i] > 100 and ep[i+1] < 10)]
-                    #print(cuts)
-                    if len(cuts) == 0:
-                        att = atta+'->'+victim+'-0'
-                        #print('---', att, [x[2] for x in episodes])
-                        keys.append(att)
-                        alerts.append(episodes)
-                        #pass
-                    else:
-                        rest = (-1,-1)
-                       
-                        for i in range(len(cuts)):
-                            start, end = 0, 0
-                            if i == 0:
-                                start = 0
-                                end = cuts[i]  
-                            else:
-                                start = cuts[i-1]+1
-                                end = cuts[i]
-                            rest = (end+1, len(ep)-1)
-                            al = episodes[start:end+1]
-                            if len(al) < 2:
-                                #print('discarding symbol ', [x[2] for x in al])#, start, end, len(episodes))
-                                continue
-                            att = atta+'->'+victim+'-'+str(c)
-                            #print('---', att, [x[2] for x in al])
-                            keys.append(att)
-                            alerts.append(al)
-                            c+=1
-                        #print('--', ep[rest[0]: rest[1]+1])
-                        al = episodes[rest[0]: rest[1]+1]
-                        if len(al) < 2:
-                            #print('discarding symbol ', [x[2] for x in al]) # TODO This one is not cool1
-                            continue
-                        att = atta+'->'+victim+'-'+str(c)
-                        #print('---', att, [x[2] for x in al])
-                        keys.append(att)
-                        alerts.append(al)
-    print('\n# sub-sequences', len(keys))
-    return (alerts, keys)
-
-## 27 Aug 2020: Generating traces for flexfringe 
-def generate_traces(alerts, keys, datafile, test_ratio=0.0):
-    victims = alerts
-    al_services = [[most_frequent(y[6]) for y in x] for x in victims]
-    print('----- All unique servcies -----')
-    print(set([item for sublist in al_services for item in sublist]))
+            # Cut episode sequence when a low-severity episode follows a high-severity episode
+            count = 0
+            mcats = [epi[2] for epi in episodes]
+            cuts = [i for i in range(len(episodes) - 1) if (len(str(mcats[i])) > len(str(mcats[i + 1])))]  # (ep[i] > 100 and ep[i+1] < 10)]
```

Comment:
This is the part I would like to discuss in more detail. During RP, I have noticed that there were many traces from the root node of the S-PDFA that were low-severity episodes, which means that episode subsequence ended with a low-severity state. This, however, should not happen because SAGE accentuates high-severity states (one of its strong points).

Firstly, looking at this line, `len(str(mcats[i])) > len(str(mcats[i + 1]))` will also be true for med-sev episode and then low-sev episode (e.g. from FF trace file: "ACE|microsoft-ds ACE|microsoft-ds exfil|microsoft-ds", ACE is med-sev and it is at the end of the episode subsequence). So, I think the version that is currently commented out is closer to the truth (or maybe even `len(str(mcats[i])) == 3 and len(str(mcats[i])) > len(str(mcats[i + 1]))`, i.e. a low- or mid-sev episode follows a high-sev episode).

Secondly, it might be the case that there is no such cut possible (e.g. the entire episode sequence consists of low- and med-sev episodes), or that the part after the last cut consists only of low- and medium severity episodes (meaning that this last part is not an episode subsequence, since it does not end with a high-severity episode). The question then arises whether such parts should be kept or whether they can be discarded.

We can also discuss it later (maybe even during a meeting), since this is not part of the refactoring and should be handled in a different PR. I am just writing it here so that I don't forget this.

Reply:
Good point! So if we look at the last alert distribution for the alert subsequences, we see that most of them end with low-sev alerts. It's likely that most attackers do probing frequently between other attack attempts. Many traces also only ever end in med-severity events. This is why of course high-sev events are rare. 

Anyhow, this part of the code only needs to cut the sequences at appropriate places -- no discarding is needed (the discarding already happens later in make_AG()). So here, the viable cuts are: [med, low], [high, low], and [high, med]. For the latter two cases, we will see the first subsequence in the AGs related to the objective = high. For the first case, unless AGs are also made for medium severity objectives, they will not be appearing anywhere (and we call them partial paths).

Reply:
I see. I think it might be documented somewhere. Since I got confused, as the paper had this line:
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/ca562cfc-9337-48af-a924-7d75325a66d4)